File: newsmodule/src/main/java/zjm/cst/dhu/newsmodule/mvp/ui/fragment/base/BaseFragment.java
package zjm.cst.dhu.newsmodule.mvp.ui.fragment.base;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.squareup.leakcanary.RefWatcher;

import zjm.cst.dhu.library.Constant;
import zjm.cst.dhu.library.MyApplication;
import zjm.cst.dhu.library.mvp.presenter.BasePresenter;
import zjm.cst.dhu.newsmodule.inject.component.DaggerFragmentComponent;
import zjm.cst.dhu.newsmodule.inject.component.FragmentComponent;
import zjm.cst.dhu.newsmodule.inject.module.FragmentModule;

/**
 * Created by zjm on 2017/5/10.
 */

public abstract class BaseFragment<T extends BasePresenter> extends Fragment {
    protected T myPresenter;
    protected FragmentComponent myFragmentComponent;

    public abstract void dataBinding(View rootView);

    public abstract void setupInjector();

    public abstract void setupViews(View rootView);

    public abstract int getFragmentId();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setupFragmentComponent();
        setupInjector();
    }

    private void setupFragmentComponent() {
        myFragmentComponent = DaggerFragmentComponent.builder()
                .applicationComponent(((MyApplication) getActivity().getApplication()).getApplicationComponent())
                .fragmentModule(new FragmentModule(this))
                .build();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        RefWatcher refWatcher = MyApplication.getRefWatcher(getActivity());
        refWatcher.watch(this);
        if (myPresenter != null) {
            myPresenter.detachView();
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(getFragmentId(), container, false);
        dataBinding(rootView);
        setupViews(rootView);
        return rootView;
    }
}

File: newsmodule/src/main/kotlin/zjm.cst.dhu.newsmodule/mvp/ui/fragment/base/BaseFragment.kt
package zjm.cst.dhu.newsmodule.mvp.ui.fragment.base

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.squareup.leakcanary.RefWatcher

import zjm.cst.dhu.library.MyApplication
import zjm.cst.dhu.library.mvp.presenter.BasePresenter
import zjm.cst.dhu.newsmodule.inject.component.DaggerFragmentComponent
import zjm.cst.dhu.newsmodule.inject.component.FragmentComponent
import zjm.cst.dhu.newsmodule.inject.module.FragmentModule

/**
 * Created by zjm on 2017/5/10.
 */

abstract class BaseFragment<T : BasePresenter<*>>
    : Fragment() {
    protected var myPresenter: T? = null
    protected var myFragmentComponent: FragmentComponent?=null

    abstract fun dataBinding(rootView: View)

    abstract fun setupInjector()

    abstract fun setupViews(rootView: View)

    abstract val fragmentId: Int

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setupFragmentComponent()
        setupInjector()
    }

    private fun setupFragmentComponent() {
        myFragmentComponent = DaggerFragmentComponent.builder()
                .applicationComponent((activity.application as MyApplication).applicationComponent)
                .fragmentModule(FragmentModule(this))
                .build()
    }

    override fun onDestroy() {
        super.onDestroy()
        val refWatcher = MyApplication.getRefWatcher(activity)
        refWatcher.watch(this)
        if (myPresenter != null) {
            myPresenter!!.detachView()
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?):
            View? {
        val rootView = inflater!!.inflate(fragmentId, container, false)
        dataBinding(rootView)
        setupViews(rootView)
        return rootView
    }
}
